[AArch64] Use 0-cycle reg2reg MOVs for FPR32, FPR16, FPR8

tomershafir opened the pull request:
This change emits optimized copy instructions for FPR32, FPR16, FPR8 register classes on targets that support it. The implementation is similar to what has been done for GPR32. It adds 2 regression tests for FPR32 and FPR16.

Depends on: https://github.com/llvm/llvm-project/pull/143680 to resolve the test structure.

jroelofs commented on llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:
```diff
@@ -5302,30 +5302,73 @@ void AArch64InstrInfo::copyPhysReg(MachineBasicBlock &MBB,
       AArch64::FPR32RegClass.contains(SrcReg)) {
-    BuildMI(MBB, I, DL, get(AArch64::FMOVSr), DestReg)
-        .addReg(SrcReg, getKillRegState(KillSrc));
+    if (Subtarget.isTargetDarwin() && Subtarget.hasZeroCycleRegMove()) {
```
ISTM this should be a new subtarget feature, maybe `hasZeroCycleFPRMove()` or something similar. It's weird to conflate per-cpu hardware tuning details with platform details like `isTargetDarwin()`.

davemgreen replied:
The zero-latency instructions on the Arm side are listed in the SWOG's and are often a little different. As far as I could tell from the code, the existing HasZeroCycleRegMove might better be described as "HasZeroCycleXRegMoveButNotWRegMove". We tend to have both if we have any, so don't prefer the xreg version over the wreg. For FPR regs it looks like if we have D we have S as free too.

My point - what do you think of representing it as the individual instructions that are expected to be free? So something like `if (Subtarget.hasZeroCycleDRegMove() && !Subtarget.hasZeroCycleSRegMove()) ...`. Hopefully that is still simple enough and not too over-engineered.

tomershafir committed:
[AArch64] Add 0-cycle copy subtarget features for FPR64, FPR32 reg classes

This change adds 2 new subtarget features to AArch64 to model 0-cycle copy execution for FPR64 and FPR32 register classes. It also adds the new `HasZeroCycleRegMovFPR64` feature to Apple processors.

tomershafir committed:
[AArch64] Use 0-cycle copy for FPR32, FPR16, FPR8

This change emits zero cycle copy instructions for FPR32, FPR16, FPR8 register classes on targets that support it. The implementation is similar to what has been done for GPR32. It adds a regression test with 2 variants for FPR32 and FPR16.

tomershafir commented:
v2:

commit 1: bb47589b1b21f0bd74ff14e47c9328114fb305e0
* Add 2 subtarget features for FPR64, FPR32 zero cycle copy, and add `FeatureZCRegMovFPR64` to Apple processors

commit 2: 7c0a52625d177f496f3488254fc7370c72c664f3
* Emit 0-cycle instructions where possible
* Add a single regression test with 2 internal function variants
* Expand `-march=arm64` test case into `-mtriple=arm64-linux-gnu` and `-mtriple=arm64-apple-macosx -mcpu=generic`

I will fix the existing `HasZeroCycleRegMov` in a followup patch.

tomershafir commented:
@jroelofs @davemgreen Ping

davemgreen reviewed (approved):
Thanks for adding the two new features. LGTM if there are not other comments from @jroelofs

jroelofs reviewed (approved):
One small nit, but otherwise LGTM.
